jax_distribution/systems/ippo_ff_anakin.py

```diff
@@ -0,0 +1,551 @@
+            batch = jax.tree_util.tree_map(
+                lambda x: x.reshape((batch_size,) + x.shape[1:]), batch
+            )
+            shuffled_batch = jax.tree_util.tree_map(
+                lambda x: jnp.take(x, permutation, axis=0), batch
+            )
+            minibatches = jax.tree_util.tree_map(
+                lambda x: jnp.reshape(
+                    x, [config["NUM_MINIBATCHES"], -1] + list(x.shape[1:])
+                ),
+                shuffled_batch,
+            )
+            (params, opt_state), total_loss = jax.lax.scan(
+                _update_minbatch, (params, opt_state), minibatches
+            )
+
+            update_state = ((params, opt_state), traj_batch, advantages, targets, rng)
+            return update_state, total_loss
+
+        update_state = ((params, opt_state), traj_batch, advantages, targets, rng)
+
+        update_state, loss_info = jax.lax.scan(
+            _update_epoch, update_state, None, config["UPDATE_EPOCHS"]
+        )
+
+        ((params, opt_state), traj_batch, advantages, targets, rng) = update_state
+
+        metric_info = jax.tree_util.tree_map(lambda x: x[:, 0], traj_batch.info)
+        return params, opt_state, rng, env_state, last_timestep, metric_info
+
+    def update_fn(
+        params: FrozenDict,
+        opt_state: optax.OptState,
+        rng: chex.PRNGKey,
+        env_state: LogEnvState,
+        timestep: TimeStep,
+    ) -> Tuple:
+        """Compute a gradient update from a single trajectory."""
+        rng, loss_rng = jax.random.split(rng)
+        (
+            new_params,
+            new_opt_state,
+            rng,
+            new_env_state,
+            new_timestep,
+            metric_info,
+        ) = update_step_fn(
+            params,
+            opt_state,
+            rng,
+            env_state,
+            timestep,
+        )
+
+        return (
+            new_params,
+            new_opt_state,
+            rng,
+            new_env_state,
+            new_timestep,
+        ), metric_info
+
+    def learner_fn(
+        params: FrozenDict,
+        opt_state: optax.OptState,
+        rngs: chex.Array,
+        env_states: LogEnvState,
+        timesteps: TimeStep,
+    ) -> Tuple:
+        """Vectorise and repeat the update."""
+        batched_update_fn = jax.vmap(
+            update_fn, axis_name="j"
+        )  # vectorize across batch.
+
+        def iterate_fn(val: Tuple, unused: Any) -> Tuple:
+            """Repeat the update function."""
+            params, opt_state, rngs, env_states, timesteps = val
+            return batched_update_fn(params, opt_state, rngs, env_states, timesteps)
+
+        runner_state = (params, opt_state, rngs, env_states, timesteps)
+        runner_state, metric = jax.lax.scan(
+            iterate_fn, runner_state, None, config["ITERATIONS"]
+        )
+        return {"runner_state": runner_state, "metrics": metric}
+
+    return learner_fn
+
+
+def run_experiment(env_name, config):
```

Comment:
We need to add docs and the inputs-output types.

Reply:
Agreed. This has now been updated :+1: